pkg/morph/deploy/contracts.go

```diff
@@ -1,7 +1,410 @@
 package deploy
 
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient/actor"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient/management"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient/nns"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/manifest"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract/nef"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/wallet"
+	"go.uber.org/zap"
+)
+
 // various common methods of the NeoFS contracts.
 const (
 	methodUpdate  = "update"
 	methodVersion = "version"
 )
+
+// syncNeoFSContractPrm groups parameters of syncNeoFSContract.
+type syncNeoFSContractPrm struct {
+	logger *zap.Logger
+
+	blockchain Blockchain
+
+	// based on blockchain
+	monitor *blockchainMonitor
+
+	localAcc *wallet.Account
+
+	// address of the NeoFS NNS contract deployed in the blockchain
+	nnsContract util.Uint160
+	systemEmail string
+
+	committee         keys.PublicKeys
+	committeeGroupKey *keys.PrivateKey
+
+	localNEF      nef.File
+	localManifest manifest.Manifest
+
+	// L2 domain name in domainContractAddresses TLD in the NNS
+	domainName string
+
+	// if set, syncNeoFSContract attempts to deploy the contract when it's
+	// missing on the chain
+	tryDeploy bool
+	// is contract must be deployed by the committee
+	committeeDeployRequired bool
+
+	// optional list of extra arguments used for test deployment of the contract.
+	// Elements must be semantically valid (type, deploy logic), at the same time,
+	// the values can be any: they will not be used for real deployment.
+	anyValidExtraDeployArgs []interface{}
+	// optional constructor of extra arguments to be passed into method deploying
+	// the contract. If returns both nil, no data is passed (noExtraDeployArgs can
+	// be used).
+	//
+	// Ignored if tryDeploy is unset.
+	buildExtraDeployArgs func() ([]interface{}, error)
+
+	// constructor of extra arguments to be passed into method updating the
+	// contract. If returns both nil, no data is passed.
+	buildVersionedExtraUpdateArgs func(versionOnChain contractVersion) ([]interface{}, error)
+}
+
+// syncNeoFSContract behaves similar to updateNNSContract but also attempts to
+// deploy the contract if it is missing on the chain and tryDeploy flag is set.
+// If committeeDeployRequired is set, the contract is deployed on behalf of the
+// committee.
+//
+// Returns address of the on-chain contract synchronized with the record of the
+// NNS domain with parameterized name.
+func syncNeoFSContract(ctx context.Context, prm syncNeoFSContractPrm) (util.Uint160, error) {
```

Comment:
300+ lines long func, up to 7 level nesting. can it be simplified?

Reply:
dont wanna do this right now, otherwise there will be more funcs called once accepting a lot of parameters

Reply:
have nothing agains synthesize funcs that are well documented and their purpose is more than just to reduce the size of other functions. But, at the same time, i prefer to do refactor when behavior is approved and stable, and everything is well-tested

Reply:
and tbh this func isnt too big: there are a lot of logs+whitespace, all actions are done in `for` loop and there aint many of them. Ofc linter says function is ugly, but any fella will figure it out i bet

Reply:
> up to 7 level nesting

lie, up to 4-5

Reply:
> behavior is approved and stable

it is hard to approve smth that hard to keep in mind

> any fella will figure it out i bet

any code (that can be compiled) can be understood. that is the reading/time question

> lie, up to 4-5

you pushed code 7+ times since, why the numbers are so necessary? "lie"? 6 [tabs](https://github.com/nspcc-dev/neofs-node/commit/e8b560df810e0d25c7ce72c0a67e79f4e7e3f865#diff-621a29555264aa6c3872e4499a659226678ec26b5b2ef9eff5b54d2dcf3fdc0bR363), the behavior changing keyword (`continue`) is the seventh in the func context